Write "kubectl options" help message to stdout, not stderr

alexandercampbell opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

Fix a very minor issue causing `kubectl` to write its help messages to `stderr` instead of `stdout`.

Try this:

`kubectl options | grep log`

It should print only the options related to logging, but right now it prints the entire help menu (since it's printing to stderr).

This patch brings us closer to unix convention and reduces user friction.

~~Another use case (if a user can't remember whether it's `-r` or `-R` for recursion):~~

~~`kubectl patch -h | grep recursive`~~

Update: this patch only affects `kubectl options`. The other commands are working as intended.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

ericchiang commented:
This redirects all error messages to stderr, not just the help.

```
$ kubectl get --not-a-flag > /dev/null 
$ 
```

alexandercampbell commented:
@ericchiang good catch. Closing to fix. Will reopen when that's resolved.

alexandercampbell commented:
@ericchiang I've done some testing with the Cobra library that we're using (specifically, the `cobra/cobra` example command) and their solution is to [catch all errors](https://github.com/spf13/cobra/blob/ca57f0f5dba473a8a58765d16d7e811fb8027add/cobra/cmd/root.go#L39-L41) and [double-print them](https://github.com/spf13/cobra/blob/ca57f0f5dba473a8a58765d16d7e811fb8027add/cobra/cmd/helpers.go#L42-L45) to both stderr and stdout.

Example incorrect command name:
```bash
[campbellalex:...ithub.com/spf13/cobra/cobra]$ ./cobra not-a-command                                                                                                        (master) 
Error: unknown command "not-a-command" for "cobra"
Run 'cobra --help' for usage.
Error: unknown command "not-a-command" for "cobra"
```

One of the `Error:` lines is stderr, the other is stdout.

The implementation is slightly different for incorrect flag names. The error is still piped to both stderr and stdout but the entire command's usage is also printed to stderr.

Would you agree that the ideal behavior is as follows (with no double-printing)?

1. Every error message &rarr; stderr.
2. Every  `"--help"` message &rarr; stdout.

If so, I will work to implement this behavior in kubectl.

ericchiang commented:
> Would you agree that the ideal behavior is as follows (with no double-printing)?

> Every error message → stderr.
> Every "--help" message → stdout.

> If so, I will work to implement this behavior in kubectl.

This actually looks like the current behavior

```
$ kubectl --help > /dev/null
$ kubectl patch -h | grep recursive
  -R, --recursive=false: Process the directory used in -f, --filename recursively. Useful when you want to manage related manifests organized within the same directory.
```

I don't know if this changed recently, but has been the case for a while in cobra https://github.com/spf13/cobra/pull/71

Commit:
cmd/options: `kubectl options` writes to `out` stream

Previous behavior was to write to stderr (thanks to the fallback system
in the Cobra library), which made it difficult to grep for flags.

For example:

	kubectl options | grep recursive

ericchiang commented:
@alexandercampbell options doesn't have a help flag. maybe look into why that is.

alexandercampbell commented:
@ericchiang what would you expect `kubectl options -h` to print? It's already a help command.

fabianofranz commented:
Don't use `os.Stdout` directly, pass the output writer to NewCmdOptions [here](https://github.com/kubernetes/kubernetes/blob/644a544d62fc27647abe976dfbe1a2cee1976db0/pkg/kubectl/cmd/cmd.go#L381) like we do in other commands, and then set it inside the [options cmd code](https://github.com/kubernetes/kubernetes/blob/644a544d62fc27647abe976dfbe1a2cee1976db0/pkg/kubectl/cmd/options.go) instead of the templater.

alexandercampbell commented:
@fabianofranz oops, sorry about that. Updated!

fabianofranz commented:
@alexandercampbell thanks! Mind squashing the commits?

alexandercampbell commented:
@fabianofranz squashed and rebased onto latest master. Thanks for your guidance on this PR.

fabianofranz commented:
@k8s-bot ok to test
/lgtm
/approve

fabianofranz commented:
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this
@k8s-bot pull-kubernetes-unit test this

alexandercampbell commented:
/assign @pwittrock

fabianofranz commented:
/approve
/lgtm

wojtek-t commented:
/approve
